src/goto-programs/goto_contractor.cpp

```diff
@@ -98,39 +97,49 @@ void goto_contractort::parse_intervals(irep_container<expr2t> expr)
 void goto_contractort::insert_assume(goto_functionst goto_functions)
 {
   loopst loop;
+  unsigned int last_loc = 0;
+
+  ///This loop is to find the last loop in the code based on location
   for(auto &function_loop : function_loops)
-    loop = function_loop;
+    if(last_loc < function_loop.get_original_loop_head()->location_number)
+    {
+      loop = function_loop;
+      last_loc = loop.get_original_loop_head()->location_number;
+    }
 
   auto loop_exit = loop.get_original_loop_exit();
 
   goto_programt dest;
 
   auto goto_function = goto_functions.function_map.find("c:@F@main")->second;
 
-  for(auto const &var : map.var_map)
+  if(map.is_empty_set())
```

Comment:
this is not what I meant, do you only need to check if the var_map is empty here? What do you call `map.var_map.empty()` instead?

Reply:
That would be incorrect. Because the check for an empty set is done on the temporarily created IntervalVector object by contractor(). The var_map can have members and the is_empty_set() is true.

Reply:
So, if I got it right: 
1. map holds all constraints.
2. If we didn't add any new constraints during a given pass, we don't add them to the program.
3. If there were constraints added to the map, we add them all to the program.

Is that correct? If so, it might be unsound.

Reply:
I don't understand what you mean here 100% though I'll try to answer.

> 1. map holds all constraints.

Yes, it holds each variable that is in the property and its interval. Also holds the property which is our constraint in the Constraint Satisfaction Problem to be used for the contractor. 
> 2.  If we didn't add any new constraints during a given pass, we don't add them to the program.
> 3. If there were constraints added to the map, we add them all to the program.

If the map is empty, we can't apply the contractor; thus, we do not add any `assume()` into the program. If the map is not empty, we apply the contractor and check the result: 

1. If the returned IntervalVector is an empty set, we insert `assume(0)` into the loop. Because it means that the property will never be violated.
2. Otherwise, for each variable, If the intervals have changed, insert assume reflecting the new interval.  

I hope this clarifies the idea. We could also have a small chat online because I would like to hear your thoughts about it.